flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -1237,6 +1262,47 @@ public FileVisitResult visitFile(java.nio.file.Path file, BasicFileAttributes at
 		return classPaths;
 	}
 
+	private List<String> registerSharedLibs(
+			Collection<Path> sharedLibs,
+			FileSystem fs,
+			List<Path> remotePaths,
+			Map<String, LocalResource> localResources,
+			StringBuilder envShipFileList) throws IOException {
+		List<String> classPaths = new ArrayList<>();
+
+		for (Path sharedLibPath : sharedLibs) {
+			if (fs.isFile(sharedLibPath)) {
+				String classpath = setupSharedLib(sharedLibPath, fs, remotePaths, localResources, envShipFileList);
+				classPaths.add(classpath);
+			} else { // directory
+				RemoteIterator<LocatedFileStatus> i = fs.listFiles(sharedLibPath, true);
+				while (i.hasNext()) {
+					LocatedFileStatus fileStatus = i.next();
+					if (fileStatus.isFile()) {
+						String classpath = setupSharedLib(fileStatus.getPath(), fs, remotePaths, localResources, envShipFileList);
```

Comment:
Here we flatten all files in the root relative path of program. Maybe it is better we keep the directory layout. Locally changed and verifying...